snark/src/r1cs.rs

```diff
@@ -0,0 +1,70 @@
+use ark_relations::r1cs::{R1CS, ConstraintSynthesizer};
+
+/// The basic functionality for a SNARK.
+pub trait SNARKForR1CS<F: Field>: SNARK<R1CS<F>> {
+    /// Generates a proof of satisfaction of the arithmetic circuit C (specified
+    /// as R1CS constraints).
+    fn prove<C: ConstraintSynthesizer<F>, Rng: RngCore + CryptoRng>(
+        circuit_pk: &Self::ProvingKey,
+        witness_generator: C,
+        rng: &mut Rng,
+    ) -> Result<Self::Proof, Self::Error>;
+
+    /// Checks that `proof` is a valid proof of the satisfaction of circuit
+    /// encoded in `circuit_vk`, with respect to the public input `public_input`,
+    /// specified as R1CS constraints.
+    fn verify<C: ConstraintSynthesizer<F>>(
+        circuit_vk: &Self::VerifyingKey,
+        public_input_generator: C,
+        proof: &Self::Proof,
+    ) -> Result<bool, Self::Error> {
+        let pvk = Self::process_vk(circuit_vk)?;
+        Self::verify_with_processed_vk(&pvk, public_input_generator, proof)
+    }
+
+    /// Preprocesses `circuit_vk` to enable faster verification.
+    fn process_vk(
+        circuit_vk: &Self::VerifyingKey,
+    ) -> Result<Self::ProcessedVerifyingKey, Self::Error>;
+
+    /// Checks that `proof` is a valid proof of the satisfaction of circuit
+    /// encoded in `circuit_pvk`, with respect to the public input `public_input`,
+    /// specified as R1CS constraints.
+    fn verify_with_processed_vk<C: ConstraintSynthesizer<F>>(
+        circuit_pvk: &Self::ProcessedVerifyingKey,
+        public_input_generator: C,
+        proof: &Self::Proof,
+    ) -> Result<bool, Self::Error>;
+}
+
+/// A SNARK with (only) circuit-specific setup.
+pub trait CircuitSpecificSetupSNARKForR1CS<F: Field>: CircuitSpecificSetupSNARK<R1CS<F>> + SNARK<R1CS<F>> {
+    /// The setup algorithm for circuit-specific SNARKs.
+    fn setup<C: ConstraintSynthesizer<F>, Rng: RngCore + CryptoRng>(
```

Comment:
Should we have a more specific name like `circuit_specific_setup`?

Reply:
Yeah that makes sense, I’ll do that